force changing the cluster data member energy with the non linearity …

gconesab opened the pull request:
…correction, if requested, not done by default

alibuild commented:
959668dd1e60a8d7d1fa8f74273dc7dc2b60696f: approval required: 1 of @raymondEhlers (Raymond James Ehlers), @loizides (Constantinos Loizides), @gconesab (Gustavo Conesa Balbastre), @amatyja (Adam Tomasz Matyja), @jdmulligan (James Mulligan), @efranzis (Eliane Epple), @mfasDa (Markus Fasel), @aiola (Salvatore Aiola), @germainma (Marie Germain)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliPhysics/pull/5458) and comment at the bottom of the page.**_

gconesab commented:
please @raymondEhlers  and @jdmulligan  check if you agree

jdmulligan commented:
Hi @gconesab ,
What is the reason for this feature? In general I would be inclined to avoid adding special cases like this. I think we have a clear, documented scheme of how to store the cluster energy at various stages of correction, and this may add confusion and bulk (especially over time, when more and more special cases are added). But if it is a big headache for you otherwise, I think we can also be practical and live with it.

gconesab commented:
Hi @jdmulligan ,
I have expressed several times my discomfort with the current approach. I think the cluster->E() method should return the energy after all corrections applied and all my analysis code is based on this premise. If one wants to recover the original cluster energy, it is quite simple just summing up the energies of the cells attached to the cluster. Now I realised again when comparing my clusterization task and the correction tasks that the correction framework was not doing what I expected. I prefer to add this simple feature for those not relying on the EMCal analysis framework where I assume this is handled properly.
Let me know if you think there is another better way to do it.

jdmulligan commented:
Hi @gconesab ,

Okay, then I won't object, I have no (non-disruptive) better way. Ideally we would settle on a single convention and document it (I would be fine with either convention), but that's the usual problem all over Aliphysics. To be honest I think it would probably be less confusing if in the current approach we called it e.g. RawE() instead of E(), to avoid such confusion. But go ahead from my side, I think however Raymond has a few minor technical comments.

Best,
James

raymondEhlers commented on PWG/EMCAL/EMCALtasks/AliEmcalCorrectionClusterNonLinearity.cxx:
```diff
@@ -39,8 +39,8 @@ AliEmcalCorrectionClusterNonLinearity::AliEmcalCorrectionClusterNonLinearity() :
-  fEnergyTimeHistAfter(0)
-
+  fEnergyTimeHistAfter(0),
+  fSetClusterE(0)
```
Please init with `kFALSE` so that we don't have to rely on implicit conversions (not that this would cause a problem here, but best to be consistent)

raymondEhlers commented on PWG/EMCAL/config/AliEmcalCorrectionConfiguration.yaml:
```diff
@@ -149,6 +149,7 @@ ClusterNonLinearity:                                # Cluster non-linearity corr
     enabled: false                                  # Whether to enable the task
     createHistos: false                             # Whether the task should create output histograms
     nonLinFunct: kBeamTestCorrectedv3               # Sets the non-linearity function. Possible options are defined in the header of the cluster non-linearity correction component
+    setClusterE: false                              # Modify the cluster energy data member
```
We suggest naming this `setForceClusterE`, and I would suggest making this clear that it is for backwards compatibility and not generally recommended. Perhaps the comment could be:

```
# Only for backwards compatibility, force cluster->E() to be set to the cluster non-linearity corrected energy. Off by default. For the standard methods, see: http://alidoc.cern.ch/AliPhysics/master/READMEcontainers.html#emcalContainerClusterEnergyCorrections
```

raymondEhlers commented on PWG/EMCAL/EMCALtasks/AliEmcalCorrectionClusterNonLinearity.h:
```diff
@@ -43,6 +43,8 @@ class AliEmcalCorrectionClusterNonLinearity : public AliEmcalCorrectionComponent
   TH1F                  *fEnergyDistAfter;           //!<!energy distribution after
   TH2F                  *fEnergyTimeHistAfter;       //!<!energy/time distribution after
   
+  Bool_t                 fSetClusterE;               ///< Stored it in cluster->E() and not cluster->GetNonLinCorrEnergy()
```
Note that this comment isn't quite right, because we still set the cluster non-linearity energy regardless of the setting.

Anyway, in conjunction with the rename in the YAML config, I would suggest that this should be `fForceSetClusterE` and the comment should be along the lines of

```
// Only for backwards compatibility, force cluster->E() to be set to the cluster non-linearity corrected energy. Off by default. For the standard methods, see: http://alidoc.cern.ch/AliPhysics/master/READMEcontainers.html#emcalContainerClusterEnergyCorrections
```

alibuild commented:
61b084ca57b88c35e67d069be5082c432f7a7a52: approval required: 1 of @raymondEhlers (Raymond James Ehlers), @loizides (Constantinos Loizides), @gconesab (Gustavo Conesa Balbastre), @amatyja (Adam Tomasz Matyja), @jdmulligan (James Mulligan), @efranzis (Eliane Epple), @mfasDa (Markus Fasel), @aiola (Salvatore Aiola), @germainma (Marie Germain)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliPhysics/pull/5458) and comment at the bottom of the page.**_

gconesab commented:
Thanks @raymondEhlers and @jdmulligan 
I have implemented your comments, plus I changed the before the correction histogram, in case of forcing it checks the E() not the other method. If not agreed, I could implement the equivalent histograms. Let me know.

raymondEhlers commented on PWG/EMCAL/EMCALtasks/AliEmcalCorrectionClusterNonLinearity.h:
```diff
@@ -43,6 +43,8 @@ class AliEmcalCorrectionClusterNonLinearity : public AliEmcalCorrectionComponent
   TH1F                  *fEnergyDistAfter;           //!<!energy distribution after
   TH2F                  *fEnergyTimeHistAfter;       //!<!energy/time distribution after
   
+  Bool_t                 fSetForceClusterE;          ///< // Only for backwards compatibility, force cluster->E() to be set to the cluster non-linearity corrected energy. Off by default. For the standard methods, see: http://alidoc.cern.ch/AliPhysics/master/READMEcontainers.html#emcalContainerClusterEnergyCorrections
```
This is super trivial, but this has a double comment marker (`///< //`) - could you please remove the `//`? I know this is trivial, but I'd like to keep things as clean as possible

jdmulligan commented:
Hi @gconesab ,
Thanks. No objection from me, but aren't these energies exactly the same value at that point?

alibuild commented:
cf45fcf47fe58048c2ec4100b4446d02e6807426: approval required: 1 of @raymondEhlers (Raymond James Ehlers), @loizides (Constantinos Loizides), @gconesab (Gustavo Conesa Balbastre), @amatyja (Adam Tomasz Matyja), @jdmulligan (James Mulligan), @efranzis (Eliane Epple), @mfasDa (Markus Fasel), @aiola (Salvatore Aiola), @germainma (Marie Germain)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliPhysics/pull/5458) and comment at the bottom of the page.**_

gconesab commented:
Thanks for spotting this stupid mistake.

gconesab commented:
Hi @jdmulligan 
yes, they are, just in case, call me paranoid ...

raymondEhlers commented:
Thanks! One point here: you actually implemented it in the after hists, which seems fine to me. So there is a typo somewhere, but I'm not sure what you intended

gconesab commented:
Hi @raymondEhlers 
I want to check after the correction the method I am interested in modifying when I requested the forcing which is E(), I am not sure what typo you mean.

gconesab commented:
Ah yes, I meant after not before, just my hidden dislexia :)

gconesab commented:
+1

alibuild commented:
cf45fcf47fe58048c2ec4100b4446d02e6807426: approved: will be automatically merged on successful tests
